src/information_theory/linear_code.lean

```diff
@@ -0,0 +1,144 @@
+/-
+Copyright (c) 2022 Bolton Bailey. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bolton Bailey
+-/
+import information_theory.hamming
+import linear_algebra.linear_independent
+import linear_algebra.affine_space.affine_subspace
+import linear_algebra.finite_dimensional
+import data.set.basic
+
+/-!
+# Linear Codes
+
+This file introduces Linear codes. A linear code is a type of error-correcting code defined as a
+linear subspace of a finite-dimensional vector space.
+
+## Main Definitions
+
+* `linear_code 𝓓 F`: The type of linear codes with domain `𝓓` over field `F`
+* `reed_solomon k D` : The code consisting of all polynomials of degree `≤ k` evaluated on a
+subset `D` of the field
+-/
+
+/--
+A linear error-correcting code, defined as a subspace of the vector space of functions from a
+domain into a field.
+-/
+def linear_code (𝓓 F : Type) [fintype 𝓓] [field F] := submodule F ( 𝓓 -> F )
```

Comment:
The `fintype` assumption is unused.

Reply:
Which incidentally is to be expected - this is a problem I ran into. It's actually not really necessary that this be finite....

Reply:
Given that you're interested in matrices associated with the code, maybe it would be best to go with domain `fin n`

Reply:
Not for the general definition.

Reply:
Incidentally, you should use `hamm (𝓓 -> F)` as the type of the vectors: this is 𝓓 -> F equipped with the hamming norm as a metric.

Reply:
Yeah I tried that but `module F (hamming (λ (d : 𝓓), F))` failed to synthesize, presumably at some point I should go back and switch that.

Reply:
Hmm - it shouldn't have done, in theory. I would be grateful if you could try and replicate the issue in case it needs a patch.

Reply:
Maybe one should not hardcode `hamm` as the metric. Different types of codes use different metrics. E.g. your code may live in a matrix space, with distance the rank distance, i.e. `dist(A,B)=rank(A-B)`.

Reply:
Yes, this is another reason I have hesitated before defining codes. I think it does want to be generalised ideally. It *might* want to be specific to integer-valued distances but this might not actually be necessary.